src/yast.cpp

```diff
@@ -0,0 +1,262 @@
+                    func_def << ", ";
+            }
+            func_def << ")" << endl;
+            func_def << "\t\"\"\"" << endl;
+        }
+        func_def << "\tfrom ycp import CallYCPFunction" << endl;
+        func_def << "\treturn CallYCPFunction(\"" + ns_name + "\", \"" + function + "\", args)" << endl;
+
+        // Register function into dictionary of new module. Returns new reference - must be decremented
+        code = PyRun_String(func_def.str().c_str(), Py_single_input, g, new_module_dict);
+        Py_XDECREF(code);
+    }
+
+    // adding variables like function from ycp to module
+    for (int i = 0; i < list_variables->size(); i++) {
+        string function = list_variables->value(i)->asString()->value();
+        stringstream func_def;
+        func_def << "def " << function << "(val=None):" << endl;
+        func_def << "\tfrom ycp import GetYCPVariable, SetYCPVariable" << endl;
+        func_def << "\tif val:" << endl;
+        func_def << "\t\tSetYCPVariable(\"" + ns_name + "\", \"" + function + "\", val)" << endl;
+        func_def << "\telse:" << endl;
+        func_def << "\t\treturn GetYCPVariable(\"" + ns_name + "\", \"" + function + "\")" << endl;
+
+        // Register function into dictionary of new module. Returns new reference - must be decremented
+        code = PyRun_String(func_def.str().c_str(), Py_single_input, g, new_module_dict);
+        Py_XDECREF(code);
+    }
+
+    delete list_functions;
+    delete list_variables;
+
+    return true;
+}
+
+/**
+ * Function handles calling ycp function from python
+ * @param const string & namespace
+ * @param const string & name of function
+ * @param ... args for function
+ * @return YCPValue return result of running function
+ */
+YCPValue CallYCPFunction(const char * namespace_name, const char * function_name, YCPList args)
+{
+    YCPValue ycpArg = YCPNull ();
+	YCPValue ycpRetValue = YCPNull ();
+
+
+    // create namespace
+    Y2Namespace *ns = getNs(namespace_name);
+
+    if (ns == NULL) {
+        y2error ("Creating namespace fault.");
+        return YCPNull();
+    }
+
+    TableEntry *sym_te = ns->table ()->find (function_name);
+
+    if (sym_te == NULL) {
+        y2error ("No such symbol %s::%s", namespace_name, function_name);
+        return YCPNull();
+    }
+
+    SymbolEntryPtr sym_entry = sym_te->sentry();
+    if (sym_entry->isVariable()) {
+        y2error("Cannot execute a variable");
+        return YCPNull();
+    }
+    constFunctionTypePtr fun_type = (constFunctionTypePtr)sym_entry->type();
+    Y2Function *func_call = ns->createFunctionCall (function_name, NULL);
+
+    if (func_call == NULL) {
+        y2error ("No such function %s::%s", namespace_name, function_name);
+        return YCPNull();
+    }
+
+    for (int i = 0; i < args.size(); i++) {
+        ycpArg = args->value(i);
+
+        if (fun_type->parameterType(i)->isSymbol() && ycpArg->isString()) {
+            ycpArg = YCPSymbol(ycpArg->asString()->value());
```

Comment:
Here is the Symbol trickery.
When we have a function call that gets a string as a param, but is expecting a YCPSymbol, we convert it to a YCPSymbol.

Reply:
This of course will only work when the Symbol is one of the parameters. Where this fails is when we need a Symbol inside a list, for example.

Reply:
and that can be problem as some methods needs list of symbols e.g.

Reply:
You can pass a list of symbols like this: ycp.SomeFunction([Symbol('a'), Symbol('b'), Symbol('c')], otherparam, etc)
This is exactly how the old python code handles this.